File: plugins/_template/src/commands/runtime_rpc.py
import json
import sys
import time


class RuntimeRpcClient:
    def __init__(self):
        self._next_id = 0

    def execute_command(self, command_id, input_value=None):
        request_id = f"rpc_{int(time.time() * 1000)}_{self._next_id}"
        self._next_id += 1
        request = {
            "id": request_id,
            "method": "command.execute",
            "params": {
                "commandId": command_id,
                "input": input_value or {},
            },
        }
        print(json.dumps(request), flush=True)
        for line in sys.stdin:
            if not line.strip():
                continue
            message = json.loads(line)
            if message.get("id") != request_id:
                continue
            if message.get("ok") is False:
                error = message.get("error") or {}
                raise RuntimeError(error.get("message") or "Host command failed")
            return message.get("data")
        raise RuntimeError("Host command response stream closed")

File: plugins/_template/src/commands/runtimeRpc.js
const readline = require('node:readline');

function createRuntimeRpcClient() {
  let nextId = 0;
  const pending = new Map();
  const rl = readline.createInterface({ input: process.stdin });

  rl.on('line', (line) => {
    if (!line.trim()) return;
    let message;
    try {
      message = JSON.parse(line);
    } catch (_error) {
      return;
    }
    const request = pending.get(message.id);
    if (!request) return;
    pending.delete(message.id);
    if (message.ok === false) {
      request.reject(new Error(message.error?.message || 'Host command failed'));
    } else {
      request.resolve(message.data);
    }
  });

  return {
    executeCommand(commandId, input) {
      const id = `rpc_${Date.now().toString(36)}_${nextId++}`;
      process.stdout.write(JSON.stringify({
        id,
        method: 'command.execute',
        params: { commandId, input },
      }) + '\n');
      return new Promise((resolve, reject) => {
        pending.set(id, { resolve, reject });
      });
    },
    close() {
      rl.close();
    },
  };
}

module.exports = { createRuntimeRpcClient };
